rfcs/extended_goal_state_functionality.md

```diff
@@ -0,0 +1,59 @@
+
+Briefly, the most significant changes proposed here are as follows:
+ 1. Sensor modules may be equipped with goal state generators.
+ 2. Goal states generated by learning modules and sensor modules will be sent to a `GoalStateSelector`, a new Monty component.
+ 3. The role of the `GoalStateSelector` is to produce a single output goal state from any number of input goal states.
+ 4. Motor systems will be responsible for selecting a policy appropriate to the goal state chosen by the `GoalStateSelector`.
+ 
+**Note**: This RFC is an outgrowth of [a previous RFC](https://github.com/thousandbrainsproject/tbp.monty/pull/328). Review that document for additional context, especially with respect to the improved saccade efficiency and supporting compositional and multi-object configurations. This RFC is intentionally limited to only the architectural changes required to facilitate work on those objectives.
+
+# Implementation
+
+![Information Flow](extended_goal_state_functionality/architecture_overview.png)
+**Proposed Routing Pathways**. Learning module (LM) and sensor module (SM) derived goal states will be routed to the `GoalStateSelector`, and the `GoalStateSelector`'s output will be routed to the motor system.
+
+The `GoalStateSelector` will receive any number of goal states derived from learning and sensor modules, and it will produce at most one goal state for consumption by the motor system. There are many potential strategies for choosing a goal state, but the simplest initial approach will be to select the goal state with the largest `confidence` value. In this way, it will act exactly as the `Monty` class currently does. More complex methods are outside the scope of this RFC.
+
+Many other Monty components will require modification. Below is a non-exhaustive list of anticipated changes.
+
+### `SensorModule`
+ - Will have an optional `gsg` attribute (i.e., it will be a `GoalStateGenerator` or `None`).
+ - Will call its `gsg`'s `step_gsg` method before exiting its own `step` method (if `gsg` is not `None`).
+ - Will have a `propose_goal_state` method that returns any number of goal states. Note that this differs from the learning module's `propose_goal_state` method which returns at most 1 goal state. Returing many goal states will enable planned work on salience-based targeting methods.
+
+All of the above changes are meant to mirror existing learning-module/GSG behavior.
+
+### `Monty`
+ - Will have a `gss` attribute to hold a `GoalStateSelector`.
+ - Existing machinery that collects goal states, performs selection, and updates the motor system will be replaced. More specifically, `Monty` will collect goal states from both learning modules and sensor modules, but it will defer goal state selection to its `gss` object. The output of the `gss` object will then be used to update the motor system.
```

Comment:
```suggestion
 - Existing machinery that collects goal states, performs selection, and updates the motor system will be replaced. More specifically, `Monty` will collect goal states from both learning modules and sensor modules, but it will defer goal state selection to its `gss` object. The output of the `gss` object will then be used to update the motor system. See the for-loop associated with `best_goal_confidence` in `_pass_infos_to_motor_system` for how this is currently implemented.
```